### PR TITLE
[FE-HW-ROUTING] - Add frontend and backend routes. Add CRUD for posts, comments, tags, likes

### PR DESCRIPTION
**Pull request by Yevhenii-Bezeha**

### CODE REVIEW

**Comment by Yevhenii-Bezeha** [comment]
The first 5 commits cover the part of the routes hw. Sorry, but I added others because if you will have time I'll be very thankful to hear feedback about others commits as well, they cover CRUD part in the client-side, cover functionality like, add post, delete, add comment, edit and delete it, sort, filter, I am not sure, maybe we don't have to do it here, but logically, I added CRUD in backend, so I needed to add it also in frontend. 

But commits "add shemas" and "
backend add crud for likes, comments, tags, posts, sorting, pagination" are about backend part. They are here because I created a branch for further project development from hw-mongodb branch( I need this PR now also).

**Comment by Yevhenii-Bezeha** [comment]
https://ibb.co/2y2nYQM
https://ibb.co/7VrqD1c
https://ibb.co/S0tJmDh
https://ibb.co/xMGNJwH

File: src/app/core/models/tagCategories.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  'Question-and-Answer',
+  'History',
+  'Science',
+  'Programming',
+  'Games',
+  'Fashion',
+  'Travel',
+  'Art',
+  'Music',
+];
```
Comment (MMaksymivUa) [inline src/app/core/models/tagCategories.ts]: create enum

File: src/app/post/components/comments/comment/comment.component.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import { IComment } from '@interfaces/IPost';
+import { userId } from '@interfaces/userId';
+
+@Component({
+  selector: 'app-comment',
+  templateUrl: './comment.component.html',
+  styleUrls: ['./comment.component.scss'],
+})
+export class CommentComponent implements OnInit {
+  @Input() public comment: IComment = { message: '', postId: '' };
```
Comment (MMaksymivUa) [inline src/app/post/components/comments/comment/comment.component.ts]: model | type

File: src/app/post/components/comments/comment/comment.component.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  @Output() public commentSelected: EventEmitter<'delete' | 'edit'> =
```
Comment (MMaksymivUa) [inline src/app/post/components/comments/comment/comment.component.ts:12]: add `type` for that

File: src/app/app.component.ts [hunk 4]
```diff
@@ -11,33 +12,34 @@ import { SidebarService } from '@services/sidebar.service';
...
-  private subSidenav: Subscription;
+  public isSideNavOpened: boolean = false;
+  public error: string = '';
+  private _subSidenav: Subscription;
 
   constructor(
-    private _modalService: ModalService,
-    private _sidebarService: SidebarService
+    private _sidebarService: SidebarService,
+    private _subjectsService: PostsSubjectsService,
```
Comment (MMaksymivUa) [inline src/app/app.component.ts]: bad naming of Service

File: src/app/app.component.ts [hunk 5]
```diff
@@ -11,33 +12,34 @@ import { SidebarService } from '@services/sidebar.service';
...
-    this.isSideNavOpened = this._sidebarService.isSidenavOpened;
-    this.subSidenav = this._sidebarService.isSidenavChanged$.subscribe(
-      (isSideNavOpened) => {
+  ngOnInit(): void {
+    this._subSidenav = this._sidebarService.isSidenavChanged$.subscribe(
+      (isSideNavOpened: boolean) => {
         this.isSideNavOpened = isSideNavOpened;
       }
     );
+    this._subjectsService._error.subscribe((error: any) => {
```
Comment (MMaksymivUa) [inline src/app/app.component.ts]: why the error has to be underscored? It is a public property.
Please, do not use underscore notation in Typescript, there are property modifiers

File: src/app/core/models/emptyPost.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  _id: '',
+  title: '',
+  description: '',
+  userId: '',
+  user: [],
+  comments: [],
+  likes: [],
+  tags: [],
+  createdAt: new Date(),
+  updatedAt: new Date(),
```
Comment (MMaksymivUa) [inline src/app/core/models/emptyPost.ts:13]: maybe it better to define it as `null`

File: src/app/core/services/likes.service.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  constructor(
+    private _http: HttpClient,
+    private _subjects: PostsSubjectsService
+  ) {}
+
+  isUserLiked(likes: ILike[]): ILike | undefined {
+    return likes.find((el: ILike) => el.userId === userId);
+  }
+
+  createLike(like: ILike) {
```
Comment (MMaksymivUa) [inline src/app/core/services/likes.service.ts]: bad naming

File: src/app/core/services/postsSubjects.service.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { Injectable } from '@angular/core';
+import { Subject } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class PostsSubjectsService {
+  public _error = new Subject<any>();
+  public _inputChanged = new Subject<void>();
```
Comment (MMaksymivUa) [inline src/app/core/services/postsSubjects.service.ts]: no underscores, please

File: src/app/post/components/post-item/post-item.component.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    this.getPost();
+    this._subInpChanged = this._subjectsService._inputChanged.subscribe(() =>
+      this.getPost()
+    );
+  }
+
+  ngOnDestroy(): void {
+    this._subRoute.unsubscribe();
+    this._subGet.unsubscribe();
+    this._subInpChanged.unsubscribe();
```
Comment (MMaksymivUa) [inline src/app/post/components/post-item/post-item.component.ts]: use one subject,
```
sub: Subject
```
``this.sub.add = this.someObservable``

File: src/app/post/components/post-item/post-item.component.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import { emptyPost } from '@interfaces/emptyPost';
+import { PostsSubjectsService } from '@services/postsSubjects.service';
+import { userId } from '@interfaces/userId';
+
+@Component({
+  selector: 'app-post-item',
+  templateUrl: './post-item.component.html',
+  styleUrls: ['./post-item.component.scss'],
+})
+export class PostItemComponent implements OnInit, OnDestroy {
```
Comment (MMaksymivUa) [inline src/app/post/components/post-item/post-item.component.ts]: too much logic in dummy PostItemComponent, add it to list

File: src/app/post/services/comments.service.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      )
+      .pipe(map((data: IResCreateComment) => data.data))
+      .subscribe({
+        next: (comment: IComment) => {
+          this._subjectsService._inputChanged.next();
+        },
+        error: (error) => {
+          this._subjectsService._error.next(error);
+        },
+      });
```
Comment (MMaksymivUa) [inline src/app/post/services/comments.service.ts]: use `tap, catchError` operators, do not subscribe in service

**Pull request merged**